packages/core/src/api/nodeUtil.ts

```diff
@@ -36,3 +53,204 @@ export function getNodeById(
+    source = { type: "drop" };
+  } else if (transaction.getMeta("history$")) {
+    source = {
+      type: transaction.getMeta("history$").redo ? "redo" : "undo",
+    };
+  } else if (transaction.getMeta("y-sync$")) {
+    source = {
+      type: "yjs-remote",
+      isChangeOrigin: transaction.getMeta("y-sync$").isChangeOrigin,
+      isUndoRedoOperation: transaction.getMeta("y-sync$").isUndoRedoOperation,
+    };
+  }
+
+  const changes: BlocksChanged<BSchema, ISchema, SSchema> = [];
+  // TODO when we upgrade to Tiptap v3, we can get the appendedTransactions which would give us things like the actual inserted Block IDs.
+  // since they are appended to the transaction via the unique-id plugin
+  const combinedTransaction = combineTransactionSteps(transaction.before, [
+    transaction,
+    ...[] /*appendedTransactions*/,
+  ]);
+
+  let prevAffectedBlocks: Block<BSchema, ISchema, SSchema>[] = [];
+  let nextAffectedBlocks: Block<BSchema, ISchema, SSchema>[] = [];
+
+  getChangedRanges(combinedTransaction).forEach((range) => {
+    // All the blocks that were in the range before the transaction
+    prevAffectedBlocks = prevAffectedBlocks.concat(
+      ...findChildrenInRange(
+        combinedTransaction.before,
+        range.oldRange,
+        (node) => node.type.isInGroup("bnBlock")
+      ).map(({ node }) =>
+        nodeToBlock(
+          node,
+          editor.schema.blockSchema,
+          editor.schema.inlineContentSchema,
+          editor.schema.styleSchema,
+          editor.blockCache
+        )
+      )
+    );
+    // All the blocks that were in the range after the transaction
+    nextAffectedBlocks = nextAffectedBlocks.concat(
+      findChildrenInRange(combinedTransaction.doc, range.newRange, (node) =>
+        node.type.isInGroup("bnBlock")
+      ).map(({ node }) =>
+        nodeToBlock(
+          node,
+          editor.schema.blockSchema,
+          editor.schema.inlineContentSchema,
+          editor.schema.styleSchema,
+          editor.blockCache
+        )
+      )
+    );
+  });
+
+  // de-duplicate by block ID
+  const nextBlockIds = new Set(nextAffectedBlocks.map((block) => block.id));
+  const prevBlockIds = new Set(prevAffectedBlocks.map((block) => block.id));
+
+  // All blocks that are newly inserted (since they did not exist in the previous state)
+  const addedBlockIds = Array.from(nextBlockIds).filter(
+    (id) => !prevBlockIds.has(id)
+  );
+
+  addedBlockIds.forEach((blockId) => {
+    changes.push({
+      type: "insert",
+      block: nextAffectedBlocks.find((block) => block.id === blockId)!,
+      source,
+      prevBlock: undefined,
+    });
+  });
+
+  // All blocks that are newly removed (since they did not exist in the next state)
+  const removedBlockIds = Array.from(prevBlockIds).filter(
+    (id) => !nextBlockIds.has(id)
+  );
+
+  removedBlockIds.forEach((blockId) => {
+    changes.push({
+      type: "delete",
+      block: prevAffectedBlocks.find((block) => block.id === blockId)!,
```

Comment:
Shouldn't `prevBlock` be set instead of `block`? Since the block previously existed and no longer does

Reply:
I thought of doing that too. But felt that if someone just wanted, "list of all changed blocks", it'd be just `getChanges().map(c=>c.block)`, whereas branching on the type is more cumbersome.

I think an argument can be made for either way (i.e. the type is 'deleted' so obviously `block` is the deleted block)

Reply:
Yep fair, I don't have a strong opinion one way or another, just wanted to make sure it was intentional